### PR TITLE
Move to pyproject.toml according to PEP 621 and use setuptools_scm

### PR DESCRIPTION
**Pull request by htorianik**
Based on #98 by @KOLANICH. Please, take a look on it

Differences:

- Fix .gitignore for setuptools_scm generated version.
- Omit test coverage of the script (`__main__.py`) as it was moved to the source dir.
- Update and restrict tox to `>=4.0.9,<5`.
- Make process of building wheel and sdist more trivial: (1) use default behavior of `build`, without any flags. (2) Let build install all build deps from pyproject.toml as it's a direct use-case of a build-frontend (PEP 517).
- Tag the last commit ([a5842b1](https://github.com/amplify-education/python-hcl2/pull/107/commits/a5842b13954f411d0b83181d49458f661409427d)) as "v4.0.1", so setuptools_scm will build it as a new version.
- Updated README, CHANGELOG and classifiers in pyproject.toml.

### CODE REVIEW

**Comment by KOLANICH** [comment]
>-/hcl2/version.py
>+hcl2/version.py
>change from absolute to relative

`.gitignore` has only relative paths, and they start from a slash. Paths not starting from a slash are not quite paths:

>If there is a separator at the beginning or middle (or both) of the pattern, then the pattern is relative to the directory level of the particular .gitignore file itself. Otherwise the pattern may also match at any level below the .gitignore level.

Read https://git-scm.com/docs/gitignore

**Comment by KOLANICH** [comment]
I guess it may be OK to just rebase removing the pair of commits doing nothing. Some other paths in `.gitignore` may also need fixing.

File: .github/workflows/test-build-publish.yml [hunk 1]
```diff
@@ -38,12 +40,14 @@ jobs:
...
         - name: Generate Lark Parser
           run: |
             pip install -r requirements.pip -e .
             python hcl2/parser.py
         - name: Build tarball
-          run: python setup.py sdist
+          run: python3 -m build
         - name: Publish to Test PyPI
           uses: pypa/gh-action-pypi-publish@master
           with:
```
Comment (rddimon) [inline .github/workflows/test-build-publish.yml:53]: We might need to split GitHub actions into 2 different files
1. First one for PR checks
2. Second for publishing on release

File: README.md [hunk 2]
```diff
@@ -43,7 +43,7 @@ with open('foo.tf', 'r') as file:
 
 ## Building From Source
 
-For development, `tox>=2.9.1` is recommended.
+For development, `tox>=4.0.9` is recommended.
 
 ### Running Tests
```
Comment (rddimon) [inline README.md:49]: Please add the `Publishing process` description here

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        - name: Set up Python ${{ matrix.python-version }}
+          uses: actions/setup-python@v4
+          with:
+            python-version: ${{ matrix.python-version }}
+        - name: Install dependencies
+          run: python -m pip install --upgrade "tox>=4.0.9,<5" build
+        - name: Build to generate the version file
+          run: python3 -m build
+        - name: Test with Tox
+          run: tox
```
Comment (rddimon) [inline .github/workflows/test.yml:27]: Could you please rename the file to pr_check.yml ?

Comment (rddimon) [reply .github/workflows/test.yml:27]: Is it possible to add support for python 3.11?

Comment (htorianik) [reply .github/workflows/test.yml:27]: It's worth to try, but what about to try it in a further PR to not overload the existing one?

Comment (rddimon) [reply .github/workflows/test.yml]: This is done

**Review by rddimon — APPROVED** [review]

**Pull request merged**